designer/server/src/main/scala/pl/touk/nussknacker/ui/component/ComponentsUsageHelper.scala

```diff
@@ -25,11 +27,43 @@ object ComponentsUsageHelper {
       }
     }
 
-    processesDetails
+    @tailrec
+    def mergeFragments(fragmentsProcessesDetails: List[BaseProcessDetails[ScenarioComponentsUsages]],
```

Comment:
Why did you do this recursively? Currently we have only two levels (we don't support fragments inside fragments). Is it for the future when we get fragments nesting?

Reply:
It's for another reason. Firstly I compute all scenario component usages by splitting all processes to scenario and fragment processes. Then for each fragment process I merge its component usages into these already computed scenario usages.

Reply:
If so, it should be possible to have a semigroup for componentId -> List[NodeId], compute it for usages in scenarios, flatmap usages for fragments for each scenario and combine them. WDYT? IMO it would be easier to read, will be linear

Reply:
I think it is good idea :)

Reply:
@arkadius I refactored the code with a hope it is more readable now